addons/portal/controllers/portal.py

```diff
@@ -308,10 +308,65 @@ def account(self, redirect=None, **post):
     def on_account_update(self, values, partner):
         pass
 
+    def _get_commercial_partners_and_users(self):
+        partner = request.env.user.partner_id
+        commercial_partners = partner.commercial_partner_id.child_ids
+        portal_users = commercial_partners.filtered(lambda partner: partner.user_ids and partner.user_ids[0].has_group('base.group_portal')) \
+                                          .mapped(lambda partner: partner.user_ids[0] if partner.user_ids else partner.user_ids)
+        return commercial_partners, portal_users
+
+    def _prepare_users_accesses_values(self):
+        values = self._prepare_portal_layout_values()
+        values.update({
+            'page_name': 'users_accesses',
+            'get_message': get_message,
+        })
+        return values
+
+    @route('/my/users_accesses', type='http', auth='user', website=True, methods=['POST'])
+    def users_accesses_invitation(self, portal_user_id=None, user_email=None, **kw):
+        portal_user_id = int(portal_user_id) if isinstance(portal_user_id, str) and portal_user_id.isdecimal() else None
+        user_email = user_email.strip() if isinstance(user_email, str) else None
+
+        values = self._prepare_users_accesses_values()
+        commercial_partners, portal_users = self._get_commercial_partners_and_users()
+
+        success_message = {'success': _("Invitation sent")}
+        if portal_user_id:
+            if portal_user_id in portal_users.ids:
+                invited_partner = portal_users.browse(portal_user_id).partner_id
+                try:
+                    invited_partner.action_resend_portal_access_invitation()
+                    values['success'] = success_message
+                except UserError as e:
+                    values['errors'] = {'error': str(e)}
+            else:
+                values['errors'] = {'error': _("The partner does not exists or you do not have the rights to reinvite them, please ask an administrator to do it for you.")}
+        elif user_email:
+            invited_partner = commercial_partners.filtered(lambda partner: partner.email == user_email)
+            if invited_partner:
+                try:
+                    invited_partner.action_grant_portal_access()
+                    values['success'] = success_message
+                    portal_users += invited_partner.user_ids[0]
+                except UserError as e:
+                    values['errors'] = {'error': str(e)}
+            else:
+                values['errors'] = {'error': _('No registered partner with the email "%(email)s" in your company, please ask an administrator to register it first.',
```

Comment:
Why not simply value['error_message'], value['success_message'], ... current values generation is a bit complicated for such a simple flow.